toolchain/sem_ir/typed_insts.h

```diff
@@ -173,6 +149,129 @@ struct AnyAggregateValue {
   InstBlockId elements_id;
 };
 
+// Common representation for various `*binding_pattern` nodes.
+struct AnyBindingPattern {
```

Comment:
Hm, I hadn't thought about also alphabetizing the categories (previously they'd been immediately before their first member), but I think this is probably appropriate. In addition, I'd be inclined to move them out to a separate file. I don't remember if we already have something in place to check that a category is compatible with all of its members, but maybe now is the time to add such a check if we don't have it already.

Reply:
OK, I've moved the categories to a separate file, and added validation logic. I'm very open to suggestions for how to make it cleaner, because my template metaprogramming skills are pretty weak.